k8s/policies.go

```diff
@@ -0,0 +1,238 @@
+	// an impossible to satisfy constraint
+	//
+	// +optional
+	Affinity *corev1.Affinity
+
+	// SecurityContext holds pod-level security attributes and common container settings.
+	// Optional: Defaults to empty.  See type description for default values of each field.
+	//
+	// each non-nil value will be merged into the function's PodSecurityContext, the value will
+	// replace any existing value or previously applied Policy
+	//
+	// +optional
+	PodSecurityContext *corev1.PodSecurityContext `json:"podSecurityContext,omitempty"`
+}
+
+// Apply adds or mutates the configuration of the Deployment with the values defined
+// in the Policy. Policies are not merged, if two policies are applied, the last Policy will
+// override preceding Policies with overlapping configurations.
+func (p Policy) Apply(deployment *appsv1.Deployment) *appsv1.Deployment {
+	if len(p.Tolerations) > 0 {
+		deployment.Spec.Template.Spec.Tolerations = append(deployment.Spec.Template.Spec.Tolerations, p.Tolerations...)
+	}
+
+	if p.RuntimeClassName != nil {
+		deployment.Spec.Template.Spec.RuntimeClassName = p.RuntimeClassName
+	}
+
+	if p.Affinity != nil {
+		// use a replacement strategy because it is not clear that merging affinities will
+		// actually produce a meaning Affinity definition, it would likely result in
+		// an impossible to satisfy constraint
+		deployment.Spec.Template.Spec.Affinity = p.Affinity
+	}
+
+	if p.PodSecurityContext != nil {
+		if deployment.Spec.Template.Spec.SecurityContext == nil {
+			deployment.Spec.Template.Spec.SecurityContext = &corev1.PodSecurityContext{}
+		}
+
+		p.PodSecurityContext.DeepCopyInto(deployment.Spec.Template.Spec.SecurityContext)
+	}
+
+	return deployment
+}
+
+// Remove is the inverse of Apply, removing the mutations that the Policy would have applied
+func (p Policy) Remove(deployment *appsv1.Deployment) *appsv1.Deployment {
+
+	for _, policyToleration := range p.Tolerations {
+		// filter the existing tolerations and then update the deployment
+		// filter without allocation implementation from
+		// https://github.com/golang/go/wiki/SliceTricks#filtering-without-allocating
+		newTolerations := deployment.Spec.Template.Spec.Tolerations[:0]
+		for _, toleration := range deployment.Spec.Template.Spec.Tolerations {
+			if !reflect.DeepEqual(policyToleration, toleration) {
+				newTolerations = append(newTolerations, toleration)
+			}
+		}
+		deployment.Spec.Template.Spec.Tolerations = newTolerations
+	}
+
+	if p.RuntimeClassName != nil {
+		if equalStrings(deployment.Spec.Template.Spec.RuntimeClassName, p.RuntimeClassName) {
+			deployment.Spec.Template.Spec.RuntimeClassName = nil
+		}
+	}
+
+	if p.Affinity != nil && reflect.DeepEqual(p.Affinity, deployment.Spec.Template.Spec.Affinity) {
+		deployment.Spec.Template.Spec.Affinity = nil
+	}
+
+	if p.PodSecurityContext != nil {
+		sc := deployment.Spec.Template.Spec.SecurityContext
+
+		if reflect.DeepEqual(p.PodSecurityContext.SELinuxOptions, sc.SELinuxOptions) {
```

Comment:
There are some options here that we are clearly not going to use or be willing to test / re-test on a regular basis - i.e. SELinux / Windows / Sysctls and so forth. I would think it's better to add them in when requested than ahead of time?

Reply:
If we want to split it, then we need to create a new security context type instead of reusing the type from the k8s api 

I am not sure what type of testing we should even promise on Policies, these are advanced pass-through settings.  We only promise that it will be passed down to the Pod.

My understanding is that Policies is for giving advanced access to the Pod definition for use cases that we can't generalize or want to include as part of the standard flow.  I think it is fair to say ["YMMV"](https://dictionary.cambridge.org/dictionary/english/ymmv) for policy fields. I don't plan to test every possible run time, but i will ensure that the run time class _is set_ on the pod